pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -129,114 +139,93 @@ type DockercfgController struct {
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("error syncing dockercfg secret %s/%s: %v", s.Namespace, s.Name, err))
 		return
 	}
-	e.enqueueServiceAccountForToken(secret)
+	// we'll need to recreate the secret
+	e.secretQueue.Add(key)
 }
 
-func (e *DockercfgController) enqueueServiceAccountForToken(tokenSecret *v1.Secret) {
+func (e *DockercfgController) enqueueServiceAccountForToken(dockerCfgSecret *v1.Secret) {
 	serviceAccount := &v1.ServiceAccount{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      tokenSecret.Annotations[v1.ServiceAccountNameKey],
-			Namespace: tokenSecret.Namespace,
-			UID:       types.UID(tokenSecret.Annotations[v1.ServiceAccountUIDKey]),
+			Name:      dockerCfgSecret.Annotations[v1.ServiceAccountNameKey],
+			Namespace: dockerCfgSecret.Namespace,
+			UID:       types.UID(dockerCfgSecret.Annotations[v1.ServiceAccountUIDKey]),
 		},
 	}
 	key, err := controller.KeyFunc(serviceAccount)
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("error syncing token secret %s/%s: %v", tokenSecret.Namespace, tokenSecret.Name, err))
+		utilruntime.HandleError(fmt.Errorf("error syncing token secret %s/%s: %v", dockerCfgSecret.Namespace, dockerCfgSecret.Name, err))
 		return
 	}
-	e.queue.Add(key)
+	e.saQueue.Add(key)
 }
 
-func (e *DockercfgController) Run(workers int, stopCh <-chan struct{}) {
+func (e *DockercfgController) Run(ctx context.Context, workers int) {
 	defer utilruntime.HandleCrash()
-	defer e.queue.ShutDown()
+	defer e.saQueue.ShutDown()
+	defer e.secretQueue.ShutDown()
 
 	klog.Infof("Starting DockercfgController controller")
 	defer klog.Infof("Shutting down DockercfgController controller")
 
 	// Wait for the store to sync before starting any work in this controller.
 	ready := make(chan struct{})
-	go e.waitForDockerURLs(ready, stopCh)
+	go e.waitForDockerURLs(ctx, ready)
 	select {
 	case <-ready:
-	case <-stopCh:
+	case <-ctx.Done():
 		return
 	}
 	klog.V(1).Infof("urls found")
 
 	// Wait for the stores to fill
-	if !cache.WaitForCacheSync(stopCh, e.serviceAccountController.HasSynced, e.secretController.HasSynced) {
+	if !cache.WaitForCacheSync(ctx.Done(), e.serviceAccountController.HasSynced, e.secretController.HasSynced) {
 		return
 	}
 	klog.V(1).Infof("caches synced")
 
 	for i := 0; i < workers; i++ {
-		go wait.Until(e.worker, time.Second, stopCh)
+		go wait.UntilWithContext(ctx, runner(e.serviceAccountWorker), time.Second)
+		go wait.UntilWithContext(ctx, runner(e.secretWorker), time.Second)
+		go wait.UntilWithContext(ctx, e.secretExpirationsChecker, ExpirationCheckPeriod)
+	}
+	<-ctx.Done()
+}
+
+func runner(syncFunc func(ctx context.Context) bool) func(ctx context.Context) {
+	return func(ctx context.Context) {
+		for syncFunc(ctx) {
+		}
 	}
-	<-stopCh
 }
 
-func (c *DockercfgController) waitForDockerURLs(ready chan<- struct{}, stopCh <-chan struct{}) {
+func (c *DockercfgController) waitForDockerURLs(ctx context.Context, ready chan<- struct{}) {
```

Comment:
separate from this PR (before or after), I wonder if these are still logically configurable.  They used to be cyclical with the rest of the controller-manager so I built this logic, but I wonder if on openshift these are now practically fixed in a way that the names can be predicted during installation and reacted to if they ever change to avoid the interlock and self-discovery logic.

Just imagining future you trying to maintain this area of code.